xarray/core/duck_array_ops.py

```diff
@@ -221,9 +222,19 @@ def asarray(data, xp=np):
 
 def as_shared_dtype(scalars_or_arrays, xp=np):
     """Cast a arrays to a shared dtype using xarray's type promotion rules."""
-    array_type_cupy = array_type("cupy")
-    if array_type_cupy and any(
-        isinstance(x, array_type_cupy) for x in scalars_or_arrays
+    if any(is_extension_array_dtype(x) for x in scalars_or_arrays):
+        extension_array_types = [
+            x.dtype for x in scalars_or_arrays if is_extension_array_dtype(x)
+        ]
+        if len(extension_array_types) == len(scalars_or_arrays) and all(
+            isinstance(x, type(extension_array_types[0])) for x in extension_array_types
+        ):
+            return scalars_or_arrays
+        raise ValueError(
+            f"Cannot cast arrays to shared type, found array types {[x.dtype for x in scalars_or_arrays]}"
+        )
+    elif array_type_cupy := array_type("cupy") and any(  # noqa: F841
+        isinstance(x, array_type_cupy) for x in scalars_or_arrays  # noqa: F821
```

Comment:
Is this kind of syntax allowed?

I suspect the CI didn't run this code:
```
$ pytest xarray/tests/test_array_api.py -x
================================= test session starts =================================
platform linux -- Python 3.10.14, pytest-8.2.0, pluggy-1.5.0
rootdir: /home/mark/git/xarray
configfile: pyproject.toml
collected 13 items

xarray/tests/test_array_api.py .F

====================================== FAILURES =======================================
__________________________________ test_aggregation ___________________________________

arrays = (<xarray.DataArray (x: 2, y: 3)> Size: 48B
array([[ 1.,  2.,  3.],
       [ 4.,  5., nan]])
Coordinates:
  * x        ...      [ 4.,  5., nan]], dtype=float64)
Coordinates:
  * x        (x) int64 16B 10 20
Dimensions without coordinates: y)

    def test_aggregation(arrays: tuple[xr.DataArray, xr.DataArray]) -> None:
        np_arr, xp_arr = arrays
>       expected = np_arr.sum()

/home/mark/git/xarray/xarray/tests/test_array_api.py:51:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
/home/mark/git/xarray/xarray/core/_aggregations.py:1857: in sum
    return self.reduce(
/home/mark/git/xarray/xarray/core/dataarray.py:3805: in reduce
    var = self.variable.reduce(func, dim, axis, keep_attrs, keepdims, **kwargs)
/home/mark/git/xarray/xarray/core/variable.py:1663: in reduce
    result = super().reduce(
/home/mark/git/xarray/xarray/namedarray/core.py:889: in reduce
    data = func(self.data, **kwargs)
/home/mark/git/xarray/xarray/core/duck_array_ops.py:427: in f
    return func(values, axis=axis, **kwargs)
/home/mark/git/xarray/xarray/core/nanops.py:99: in nansum
    result = sum_where(a, axis=axis, dtype=dtype, where=mask)
/home/mark/git/xarray/xarray/core/duck_array_ops.py:333: in sum_where
    a = where_method(xp.zeros_like(data), where, data)
/home/mark/git/xarray/xarray/core/duck_array_ops.py:349: in where_method
    return where(cond, data, other)
/home/mark/git/xarray/xarray/core/duck_array_ops.py:343: in where
    return xp.where(condition, *as_shared_dtype([x, y], xp=xp))
/home/mark/git/xarray/xarray/core/duck_array_ops.py:236: in as_shared_dtype
    elif array_type_cupy := array_type("cupy") and any(  # noqa: F841
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

.0 = <list_iterator object at 0x7e404174b6d0>

    elif array_type_cupy := array_type("cupy") and any(  # noqa: F841
>       isinstance(x, array_type_cupy) for x in scalars_or_arrays  # noqa: F821
    ):
E   NameError: free variable 'array_type_cupy' referenced before assignment in enclosing scope

/home/mark/git/xarray/xarray/core/duck_array_ops.py:237: NameError
=============================== short test summary info ===============================
FAILED xarray/tests/test_array_api.py::test_aggregation - NameError: free variable 'array_type_cupy' referenced before assignment in enclosi...
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!! stopping after 1 failures !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
============================= 1 failed, 1 passed in 0.33s =============================
```

Reply:
@hmaarrfk Interesting.  I see you fixed this.  I must have done [this when testing](https://github.com/pydata/xarray/pull/9000#issuecomment-2094913015) because I do specifically remember testing this.

Reply:
No issues. It was a straightforward fix.